File: prime_numbers/nth_prime.r
nth_prime <- function(n) {
  if (n <= 0) {
    return(NULL)
  }
  primes <- c(2)
  num <- 3
  while (length(primes) < n) {
    is_prime <- TRUE
    sqrt_num <- ceiling(sqrt(num))
    for (p in primes) {
      if (p > sqrt_num) {
        break
      }
      if (num %% p == 0) {
        is_prime <- FALSE
        break
      }
    }
    if (is_prime) {
      primes <- c(primes, num)
    }
    num <- num + 2
  }
  return(primes[length(primes)])
}

start_time <- timer()
for (i in 1:10) {
  print(nth_prime(50000))
}
cat("Dauer: ", timer() - start_time)

File: prime_numbers/nth_prime.java
import java.util.Arrays;

public class Main {
    public static void main(String[] args) {
        long start_time = System.nanoTime();
        for (int i = 0; i < 10; i++) {
            System.out.println(nth_prime(50000));
        }
        long end_time = System.nanoTime();
        System.out.printf("Dauer: %f seconds%n", (end_time - start_time) / 1_000_000_000.0);
    }

    public static int nth_prime(int n) {
        if (n <= 0) {
            return -1;
        }
        int[] primes = {2};
        int num = 3;
        while (primes.length < n) {
            boolean is_prime = true;
            int sqrt_num = (int) Math.sqrt(num) + 1;
            for (int p : primes) {
                if (p > sqrt_num) {
                    break;
                }
                if (num % p == 0) {
                    is_prime = false;
                    break;
                }
            }
            if (is_prime) {
                primes = Arrays.copyOf(primes, primes.length + 1);
                primes[primes.length - 1] = num;
            }
            num += 2;
        }
        return primes[primes.length - 1];
    }
}
